File: web3d-grok/js/hitbox.js
/* SPDX-License-Identifier: MIT
 * Copyright (c) 2026 JuniorCloud LLC and FrameForge contributors.
 * Multi-point arc hitboxes. Local only. No fetch in the tick.
 * Not a Nintendo product.
 */
export function trit(v, lo = -0.2, hi = 0.2) {
  if (v < lo) return -1;
  if (v > hi) return 1;
  return 0;
}

export function arcPoints(x, y, facing, reach = 18, lift = 10, n = 5) {
  const pts = [];
  const count = Math.max(3, n | 0);
  for (let i = 0; i < count; i++) {
    const t = i / (count - 1);
    const ang = ((-25 + 70 * t) * Math.PI) / 180;
    pts.push([x + facing * reach * Math.cos(ang), y + lift * Math.sin(ang)]);
  }
  return pts;
}

export function hits(points, vx, vy, radius = 10) {
  const r2 = radius * radius;
  for (const p of points) {
    const dx = p[0] - vx;
    const dy = p[1] - vy;
    if (dx * dx + dy * dy < r2) return true;
  }
  return false;
}

export function rangeArc(atk, vic, reach = 18) {
  if (!vic || !vic.alive) return false;
  const pts = arcPoints(atk.x, atk.y + 8, atk.facing || 1, reach, 10, 5);
  return hits(pts, vic.x, vic.y + 8, 10);
}

File: python/frameforge/hitbox.py
"""Multi-point arc hitboxes. Stdlib only. Does not write knockback."""
from __future__ import annotations
import math

def trit(v, lo=-0.2, hi=0.2):
    if v < lo:
        return -1
    if v > hi:
        return 1
    return 0


def arc_points(x, y, facing, reach=18.0, lift=10.0, n=5):
    n = max(3, int(n))
    pts = []
    for i in range(n):
        t = i / (n - 1)
        ang = math.radians(-25.0 + 70.0 * t)
        pts.append((x + facing * reach * math.cos(ang), y + lift * math.sin(ang)))
    return pts


def hits(points, vic_x, vic_y, radius=10.0) -> bool:
    r2 = radius * radius
    for px, py in points:
        dx = px - vic_x
        dy = py - vic_y
        if dx * dx + dy * dy < r2:
            return True
    return False
